### DISCUSSION TITLE
Helix doesn't need "Git integration", just use the following keymap.

### DISCUSSION THREAD

**Discussion by nik-rev** [helix-editor/helix#12045] [General]
1. install lazygit
2. add this to your configuration

```toml
[keys.normal]
C-g = [
    ":write-all", 
    ":new", 
    ":insert-output lazygit", 
    ":buffer-close!", 
    ":redraw", 
    ":reload-all"
]
```
Explanation
1. `write-all`: save all buffers so that the changes can be shown in lazygit and commited right away
2. `new`: create new empty buffer
3. `insert-output lazygit`: runs the `lazygit` command and inserts it's output into the new buffer
4. when you exit lazygit, `buffer-close!` will kill the previously empty buffer and send you pack to where you were previously
5. `redraw` is necessary to re-render the UI to show the buffers
6. `reload-all` because if e.g. you commit everything, helix will still show the symbols indicating that lines were changed / deleted on the left of the line numbers. This will prevent that

**Comment by David-Else** [comment]
On testing this I am afraid the input is messed up when trying to type into lazygit, this is me pressing the delete key:

![image](https://github.com/user-attachments/assets/e8b9fe62-0e4b-4fe6-874a-8b127fd8968e)

Now I can't exit the window.

Any suggestions? I use Kitty and this has been working perfectly:

```
[keys.normal.C-t]
l = ":sh kitty @ launch --no-response --cwd=current --type=overlay --copy-env lazygit"
```

**Reply by nik-rev** [reply]
I'm using wezterm and things work perfectly for me

**Reply by David-Else** [reply]
I used your cool reload-all idea, but it just doesn't work for me when using kitty overlays:

```
l = [ ":sh kitty @ launch --no-response --cwd=current --type=overlay --copy-env lazygit", ":reload-all" ]
```

I think it must be a bug https://github.com/helix-editor/helix/discussions/12077

**Reply by mattkang** [reply]
I'm also getting similar behaviour with `foot`. I also cannot use `<c-j>` and `<c-k>` to move commits around. Seems like something is eating modifiers

**Reply by raphaelmso** [reply]
Same thing happening with ghostty and kitty here.

Edit: Wezterm works fine

**Reply by structwafel** [reply]
@David-Else 
For kitty you can fix this by using [remote-control](https://sw.kovidgoyal.net/kitty/remote-control/).:
```
"C-g" = [
    ":write-all",
    ":run-shell-command kitty @ launch --type=overlay --cwd=current lazygit",
    ":redraw",
    ":reload-all",
]
```

Note you need [allow_remote_control](https://sw.kovidgoyal.net/kitty/conf/#opt-kitty.allow_remote_control) enabled for it to work.

**Reply by David-Else** [reply]
@structwafel Thanks for the info, but I am afraid this does not solve the reloading Helix after exiting Lazygit problem. I was already using the remote functionality like this:

```sh
l   = ":sh kitty @ launch --no-response --cwd=current --type=overlay --copy-env lazygit"
```
(`--no-response` will stop the window number return code being returned to Helix which needs dismissing every time)

Your suggestion does not actually `reload-all`, I made a thread https://github.com/helix-editor/helix/discussions/12077 about it, if you can solve it you will be truly awesome :)

**Reply by nikhilmaddirala** [reply]
Hi, I'm having the same issue with Ghostty

**Comment by ngasull** [comment]
I love what I'm reading here: combining tools that greatly do their purpose together. Thanks for sharing such ideas.

One issue I'm facing though is disabled mouse support after execution. Some open issues refer to the same problem, for example https://github.com/helix-editor/helix/issues/11853. Seems like Helix doesn't check if the sub-command stole mouse support.

In the meanwhile, here is my work around:
```toml
[keys.normal]
C-g = [
    ":write-all", 
    ":new", 
    ":insert-output lazygit", 
    ":set mouse false", # First disable mouse to hint helix into activating it
    ":set mouse true",
    ":buffer-close!", 
    ":redraw", 
    ":reload-all"
]
```

**Comment by leonlonsdale** [comment]
Working fine in Ghostty.

Would something similar work with Yazi? I guess the problem would be piping back a file to a buffer after creating & opening a new file. Of course, the  standard picker for existing files

**Reply by nouritsu** [reply]
Please continue from https://github.com/helix-editor/helix/discussions/12934

**Comment by azarmadr** [comment]
had to use `C-l` in zellij, as it conflicts with `C-g` (lock) of zellij

**Comment by Chris44442** [comment]
This doesn't work for me because after quitting Helix, the terminal doesn't get redrawn properly and is cluttered with the remnants of lazygit. Tested with Alacritty and Konsole.

**Reply by MrRobotMan** [reply]
I have a similar issue. Tested with Alacritty / powershell, Windows Terminal with powershell and cmd, CMD, and powershell. All have the same isseue where the terminal does not get redrawn after exiting Helix and all I see are the editor remnants.

**Reply by rafftre** [reply]
I'm on Windows Terminal with Powershell and using a workaround for this. I have set up an alias in C:\Users\<username>\Documents\PowerShell\Microsoft.PowerShell_profile.ps1:
```powershell
function Run-Helix-And-Clear-At-Exit {
    hx.exe $args
    Clear-Host
}
Set-Alias -Name hx -Value Run-Helix-And-Clear-At-Exit
```

**Comment by hholst80** [comment]
What does this do this spawns lazygit?  Eh, why? How would that be useful? The whole thing of an *integration* is to have it *integrated* into the editor not shell out to another tool?

**Reply by nouritsu** [reply]
Both tools are good at their own jobs. The "integration" allows you to switch to it faster than `C-z` into typing the command out.

I (and I'm sure the other participants in this thread) don't want my text editor to be shipped with an entire version control GUI. This way I get to keep my LazyGit config and also quickly spawn it into my text editor.

**Comment by mnpqraven** [comment]
There's a similar integration with yazi (a cli file manager) but with both settings I've come across the same issue where when you exit helix, the whole editor screen stays on the terminal and doesn't disappear like [this](https://github.com/sxyazi/yazi/pull/2461#issuecomment-2724443429). This happens with multiple terminal (im using both wezterm and alacritty) so i'm not sure if this is a helix issue with redrawing

another note is that if you're committing in lazygit and you select "open in editor" in the commit message window then the terminal will freeze, i have to either type out by hand or type in another terminal then copy paste the contents

**Comment by zapatoche** [comment]
For me, things are getting stucked when i try to commit.
I'm using Ghostty on Mac. In order to have Helix Meta key working, i had to remap the `macos-option-as-alt`
```
macos-option-as-alt = left
keybind = alt+left=unbind
keybind = alt+right=unbind
```
When trying to commit with the LazyGit UI (`c`) I cannot save the commit with `<Alt-enter>`.
If i try to bypass the UI with `C`, hellix freezes over.

Removing the `macos-option-as-alt` makes Helix unusable.
Maybe there's an alternative.

**Reply by zapatoche** [reply]
I've remapped <A-enter> to <C-s> and it's still not working.
I also tried to launch gitui with the same command but nothing happens.

Has anyone tried this integration with gitui?

**Reply by ndonkersloot** [reply]
> I've remapped to and it's still not working. I also tried to launch gitui with the same command but nothing happens.
> 
> Has anyone tried this integration with gitui?

This works for me:

```
C-g = [
    ":write-all", 
    ":new", 
    ":insert-output gitui >/dev/tty", 
    ":set mouse false", 
    ":set mouse true",
    ":buffer-close!", 
    ":redraw", 
    ":reload-all"
]
```

**Comment by p4vook** [comment]
Helix *does* need a VCS integration, as the whole point of having a good editor expierience is not jumping between different editors all the time. 

When you need to stage/restore individual lines/hunks (and in big projects you need to do that often!) switching to another tool for this simple task feels very awkward.

**Reply by nouritsu** [reply]
Both tools do their jobs very well, if you want a VCS integration there is always NeoVim + bloat. Besides this method is basically just a single keypress to get in and another to get out, it works 100% seamlessly for me.

**Reply by izhrs** [reply]
> Both tools do their jobs very well, if you want a VCS integration there is always NeoVim + bloat. Besides this method is basically just a single keypress to get in and another to get out, it works 100% seamlessly for me.

Even nano does the job of editing just fine. What's even your point? The idea of keeping a tool "bloat-free" is ridiculous and makes zero sense. Bloat is subjective, just because you don't use a feature doesn't mean others won't. If you don't want to use something, then don't. Why gatekeep it?

It's been over four goddamn decades since vi was launched, and we still have only three types of CLI editors:

1. Integrate ten external tools just to make it barely usable, and half of them probably won't work right
2. Throw in fifty plugins to duct-tape together a usable setup, and at least one will break every other day
3. Completely unusable forever, no matter what you do

**Comment by Frestein** [comment]
Unable to get it to work with [gitu](https://github.com/altsem/gitu). Helix continues to be displayed.

**Comment by gloaysa** [comment]
Adding to this, I created three scripts to have my most common used git operations inline under g.

They allow me to have this on the config:

- Inline blame, with extra info like the commit hash and name, and the line change.
- Hunk changes under the cursor.
- Full file changes of the last commit under the cursor.

```toml
[keys.normal.space.g]
f = "changed_file_picker"
r = ":reset-diff-change"
# inline blame
b = ":run-shell-command ~/.config/helix/utils/blame_line_pretty.sh %{buffer_name} %{cursor_line}"
# full file changes of last commit under cursor
B = ':open %sh{~/.config/helix/utils/blame_file_pretty.sh %{buffer_name} %{cursor_line}}'
# inline hunk changes
h = ':run-shell-command ~/.config/helix/utils/git-hunk.sh %{buffer_name} %{cursor_line} 3' 
```

I place the following scripts in a folder inside \`\~/.config/helix/utils\`. I tried to leave them well documented.

You can find the scripts here: https://gist.github.com/gloaysa/828707f067e3bb20da18d72fa5d4963a

And how they look in the editor:
https://freeimage.host/i/KHS17at
https://freeimage.host/i/KHS1A6N

**Comment by iihmsunn** [comment]
I've been trying to use similar approach to run other arbitrary commands, such as make, but it has to wait for the whole thing to complete before it shows the output, and everything freezes until then. I guess lazygit works because it instantly returns control to the user, but it's also capable of rendering on its own without user input such is when it's showing the fetch progress. Does anyone by any chance know an easy way to wrap a "normal" long running command so that it sort of runs inside a TUI, but the TUI doesn't actually do anything except showing the command running?

edit: actually after some experiments it doesn't look like any tui will work, in fact I couldn't make anything work at all. I've had an idea to run commands in gnu screen, which works in the terminal but errors in helix, and trying to run other random tui apps like mc or nvim just locks everything up

edit 2: also I don't really need this anymore tbh. The only task that I actually needed a selection as an input for works just fine with :sh, and the rest works outside the editor

**Comment by moritzdietz** [comment]
I'm also adding to this with my version as I bet others might have a similar setup.

The setup is like this:
```text
~/Documents/Repos/basedir
├── repo-with-branches
│   └── .git
│
├── another-repo
│   └── .git
│
├── Subfolder 1
│   ├── some-repo
│   │   └── .git
│   └── another-repo
│       └── .git
│
├── Subfolder 2
│   ├── we-got-repos-for-days
│   │   └── .git
│   └── yup-we-do
│       └── .git
│
└── Another Folder
    ├── not-enough-repos
    │   └── .git
    └── but-now-for-real
        └── .git
```

Now when you `cd` to `~/Documents/Repos/basedir` and then start `hx` your cwd is at your "repo root" if you so want.
In order to have support for lazygit, you need to tell lazygit where to find the repo.

**Note**: this uses `%{file_path_absolute}` which is only available on main

```toml
[keys.normal]
C-g = [
    ":sh echo \"%{file_path_absolute}\" > ~/.cache/helix/file_path_absolute",
    ":write-all", 
    ":new", 
    ":insert-output lazygit -p \"%sh{git -C ''%sh{dirname %sh{cat ~/.cache/helix/file_path_absolute}}'' rev-parse --show-toplevel}\"", 
    ":buffer-close!", 
    ":set mouse false", 
    ":set mouse true",
    ":redraw", 
    ":reload-all"
]
```

Here is an example for when you are inside tmux and would rather like to start lazygit inside a tmux popup for the repo of the file you currently have open.

```toml
[keys.normal]
C-k = [
    ":sh echo \"%{file_path_absolute}\" > ~/.cache/helix/file_path_absolute",
    ":sh tmux popup -E -d '#{pane_current_path}' -w 80%% -h 80%% lazygit -p \"%sh{git -C ''%sh{dirname %{file_path_absolute}}'' rev-parse --show-toplevel}\""
]
```

The keys for the keybinding are pretty random, so…
